tests/torchtune/modules/peft/test_dora.py

```diff
@@ -0,0 +1,195 @@
+            alpha=ALPHA,
+            use_bias=False,
+        )
+
+        fixed_init_model(dora_linear)
+        return dora_linear
+
+    @pytest.fixture
+    def qdora_linear(self, in_dim, out_dim) -> DoRALinear:
+        with utils.set_default_dtype(torch.bfloat16):
+            qdora_linear = DoRALinear(
+                in_dim=512,
+                out_dim=512,
+                rank=RANK,
+                alpha=ALPHA,
+                use_bias=False,
+                quantize_base=True,
+            )
+            fixed_init_model(qdora_linear, dtype=torch.bfloat16)
+            return qdora_linear
+
+    def test_forward(self, inputs, dora_linear, out_dim) -> None:
+        expected = torch.tensor(EXPECTED_VAL)
+        actual = dora_linear(inputs)
+        assert actual.shape == (BSZ, SEQ_LEN, out_dim)
+        torch.testing.assert_close(actual.mean(), expected, atol=1e-4, rtol=1e-6)
+
+    def test_dora_weight_nf4_when_quantized(self, qdora_linear):
+        assert isinstance(qdora_linear.weight, NF4Tensor)
+
+    def test_bias_raises(self):
+        with pytest.raises(
+            NotImplementedError, match="DoRALinear does not support using bias"
+        ):
+            DoRALinear(
+                in_dim=512,
+                out_dim=512,
+                rank=RANK,
+                alpha=ALPHA,
+                use_bias=True,
+                quantize_base=False,
+            )
+
+    # @pytest.mark.parametrize("dtype", [torch.bfloat16, torch.float32])
+    def test_qdora_parity(self):
+        dtype = torch.bfloat16
+        with utils.set_default_dtype(dtype):
+            torch.manual_seed(0)
+            qdora_linear = DoRALinear(
+                in_dim=512,
+                out_dim=512,
+                rank=RANK,
+                alpha=ALPHA,
+                use_bias=False,
+                quantize_base=True,
+            )
+            torch.manual_seed(0)
+            dora_linear = DoRALinear(
+                in_dim=512,
+                out_dim=512,
+                rank=RANK,
+                alpha=ALPHA,
+                use_bias=False,
+                quantize_base=False,
+            )
+
+        qdora_linear.initialize_dora_magnitude()
+        dora_linear.initialize_dora_magnitude()
+
+        # set weight of dora_linear to unquantized weight of qdora_linear and check
+        # parity.
+        dora_linear.weight.data = qdora_linear.weight.to(dtype)
+
+        # Ensure forward passes are the same. This is because DoRALinear should use a special
+        # quantized linear operator that runs compute in higher prec (but only saves the 4 bit quantized tensor)
+        # for autograd.
+        inputs = torch.randn(BSZ, SEQ_LEN, 512, dtype=dtype)
+        torch.manual_seed(0)
+        dora_linear_out = dora_linear(inputs)
+        torch.manual_seed(0)
+        qdora_linear_out = qdora_linear(inputs)
+        torch.testing.assert_close(
+            dora_linear_out, qdora_linear_out, rtol=1e-01, atol=1e-01
```

Comment:
I need to make this assert very loose since I found to_nf4 will change the original value significantly

Reply:
Hmm this is a bit suspicious to me. For comparison in the [corresponding LoRA test](https://github.com/pytorch/torchtune/blob/a83eeff0079a73ee04a11e8fc2573ed8f671b231/tests/torchtune/modules/peft/test_lora.py#L144) we do not have to do this. Can we make sure nothing unexpected is happening with the magnitude vector here?